test/jdk/sun/security/pkcs11/fips/VerifyMissingAttributes.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright (c) 2022, Red Hat, Inc.
+ *
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+import java.security.Provider;
+import java.security.Security;
+import java.util.HashMap;
+import java.util.Map;
+
+/*
+ * @test
+ * @bug 9999999
+ * @requires (jdk.version.major >= 8)
+ * @run main/othervm/timeout=30 Main
+ * @author Martin Balao (mbalao@redhat.com)
+ */
+
+public final class VerifyMissingAttributes {
+
+    private static final String[] svcAlgImplementedIn = {
+            "AlgorithmParameterGenerator.DSA",
+            "AlgorithmParameters.DSA",
+            "CertificateFactory.X.509",
+            "KeyStore.JKS",
+            "KeyStore.CaseExactJKS",
+            "KeyStore.DKS",
+            "CertStore.Collection",
+            "CertStore.com.sun.security.IndexedCollection"
+    };
+
+    public static void main(String[] args) throws Throwable {
+        Map<String, String> attrs = new HashMap<>();
+        Provider sunProvider = Security.getProvider("SUN");
+        for (String svcAlg : svcAlgImplementedIn) {
+            attrs.clear();
+            attrs.put(svcAlg + " ImplementedIn", "Software");
+            doQuery(sunProvider, attrs);
+        }
+        if (Double.parseDouble(
+                System.getProperty("java.specification.version")) >= 17) {
+            attrs.clear();
+            attrs.put("KeyFactory.RSASSA-PSS SupportedKeyClasses",
+                    "java.security.interfaces.RSAPublicKey" +
+                            "|java.security.interfaces.RSAPrivateKey");
+            doQuery(Security.getProvider("SunRsaSign"), attrs);
```

Comment:
I see that we are always calling `attrs.clear()` + `attrs.put(...)` in a way that only a single attribute is used at a time. So we don't need a `Map<String, String> attrs`, we can use [`Security.getProviders(String)`](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/security/Security.html#getProviders(java.lang.String) "java.security.Security::getProviders(java.lang.String) documentation"). The `String filter = "<crypto_service>.<algorithm_or_type> <attribute_name>:<attribute_value>"` usage seems to cover all our cases.

Later, when creating the test failure exception, this would also remove the need to iterate through `attrs` and use a `StringBuffer`. Please note that by just appending `filter` instead of `sb.toString()`, we would get almost the same text (but without a space after the colon in `<key>:<value>`).

I won't block the approval for this.

----
`Security.getProviders(String)` implementation:
https://github.com/rh-openjdk/jdk/blob/0bd5ca9ccc5890b009b927b541424d0aacbb0463/src/java.base/share/classes/java/security/Security.java#L616-L633

Reply:
I've taken this suggestion, thanks. Please have a look again.